File: networking/src/main/java/ch/dissem/bitmessage/networking/Connection.java
/*
 * Copyright 2015 Christian Basler
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package ch.dissem.bitmessage.networking;

import ch.dissem.bitmessage.InternalContext;
import ch.dissem.bitmessage.entity.GetData;
import ch.dissem.bitmessage.entity.MessagePayload;
import ch.dissem.bitmessage.entity.NetworkMessage;
import ch.dissem.bitmessage.entity.Version;
import ch.dissem.bitmessage.entity.valueobject.InventoryVector;
import ch.dissem.bitmessage.entity.valueobject.NetworkAddress;
import ch.dissem.bitmessage.factory.Factory;
import ch.dissem.bitmessage.ports.NetworkHandler.MessageListener;
import ch.dissem.bitmessage.utils.UnixTime;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.InetAddress;
import java.net.InetSocketAddress;
import java.net.Socket;
import java.net.SocketTimeoutException;
import java.util.HashSet;
import java.util.Objects;
import java.util.Set;

import static ch.dissem.bitmessage.networking.AbstractConnection.Mode.CLIENT;
import static ch.dissem.bitmessage.networking.AbstractConnection.Mode.SYNC;
import static ch.dissem.bitmessage.networking.AbstractConnection.State.ACTIVE;
import static ch.dissem.bitmessage.networking.AbstractConnection.State.DISCONNECTED;
import static ch.dissem.bitmessage.utils.UnixTime.MINUTE;

/**
 * A connection to a specific node
 */
class Connection extends AbstractConnection {
    public static final int READ_TIMEOUT = 2000;
    private static final Logger LOG = LoggerFactory.getLogger(Connection.class);
    private static final int CONNECT_TIMEOUT = 5000;

    private final long startTime;
    private final Socket socket;
    private final ReaderRunnable reader = new ReaderRunnable();
    private final WriterRunnable writer = new WriterRunnable();

    private InputStream in;
    private OutputStream out;
    private boolean socketInitialized;

    public Connection(InternalContext context, Mode mode, Socket socket, MessageListener listener,
                      Set<InventoryVector> requestedObjectsMap) throws IOException {
        this(context, mode, listener, socket, requestedObjectsMap,
                new NetworkAddress.Builder().ip(socket.getInetAddress()).port(socket.getPort()).stream(1).build(),
                0);
    }

    public Connection(InternalContext context, Mode mode, NetworkAddress node, MessageListener listener,
                      Set<InventoryVector> requestedObjectsMap) {
        this(context, mode, listener, new Socket(), requestedObjectsMap,
                node, 0);
    }

    private Connection(InternalContext context, Mode mode, MessageListener listener, Socket socket,
                       Set<InventoryVector> commonRequestedObjects, NetworkAddress node, long syncTimeout) {
        super(context, mode, node, listener, commonRequestedObjects, syncTimeout);
        this.startTime = UnixTime.now();
        this.socket = socket;
    }

    public static Connection sync(InternalContext ctx, InetAddress address, int port, MessageListener listener,
                                  long timeoutInSeconds) throws IOException {
        return new Connection(ctx, SYNC, listener, new Socket(address, port),
                new HashSet<InventoryVector>(),
                new NetworkAddress.Builder().ip(address).port(port).stream(1).build(),
                timeoutInSeconds);
    }

    public long getStartTime() {
        return startTime;
    }

    public Mode getMode() {
        return mode;
    }

    public State getState() {
        return state;
    }

    public NetworkAddress getNode() {
        return node;
    }

    @Override
    protected void send(MessagePayload payload) {
        try {
            if (payload instanceof GetData) {
                requestedObjects.addAll(((GetData) payload).getInventory());
            }
            synchronized (this) {
                new NetworkMessage(payload).write(out);
            }
        } catch (IOException e) {
            LOG.error(e.getMessage(), e);
            disconnect();
        }
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Connection that = (Connection) o;
        return Objects.equals(node, that.node);
    }

    @Override
    public int hashCode() {
        return Objects.hash(node);
    }

    private synchronized void initSocket(Socket socket) throws IOException {
        if (!socketInitialized) {
            if (!socket.isConnected()) {
                LOG.trace("Trying to connect to node " + node);
                socket.connect(new InetSocketAddress(node.toInetAddress(), node.getPort()), CONNECT_TIMEOUT);
            }
            socket.setSoTimeout(READ_TIMEOUT);
            in = socket.getInputStream();
            out = socket.getOutputStream();
            socketInitialized = true;
        }
    }

    public ReaderRunnable getReader() {
        return reader;
    }

    public WriterRunnable getWriter() {
        return writer;
    }

    public class ReaderRunnable implements Runnable {
        @Override
        public void run() {
            try (Socket socket = Connection.this.socket) {
                initSocket(socket);
                if (mode == CLIENT || mode == SYNC) {
                    send(new Version.Builder().defaults(peerNonce).addrFrom(host).addrRecv(node).build());
                }
                while (state != DISCONNECTED) {
                    if (mode != SYNC) {
                        if (state == ACTIVE && requestedObjects.isEmpty() && sendingQueue.isEmpty()) {
                            Thread.sleep(1000);
                        } else {
                            Thread.sleep(100);
                        }
                    }
                    receive();
                }
            } catch (Exception e) {
                LOG.trace("Reader disconnected from node " + node + ": " + e.getMessage());
            } finally {
                disconnect();
                try {
                    socket.close();
                } catch (Exception e) {
                    LOG.debug(e.getMessage(), e);
                }
            }
        }

        private void receive() throws InterruptedException {
            try {
                NetworkMessage msg = Factory.getNetworkMessage(version, in);
                if (msg == null)
                    return;
                handleMessage(msg.getPayload());
                if (socket.isClosed() || syncFinished(msg) || checkOpenRequests()) disconnect();
            } catch (SocketTimeoutException ignore) {
                if (state == ACTIVE && syncFinished(null)) disconnect();
            }
        }

    }

    private boolean checkOpenRequests() {
        return !requestedObjects.isEmpty() && lastObjectTime > 0 && (UnixTime.now() - lastObjectTime) > 2 * MINUTE;
    }

    public class WriterRunnable implements Runnable {
        @Override
        public void run() {
            try (Socket socket = Connection.this.socket) {
                initSocket(socket);
                while (state != DISCONNECTED) {
                    if (sendingQueue.isEmpty()) {
                        Thread.sleep(1000);
                    } else {
                        send(sendingQueue.poll());
                    }
                }
            } catch (IOException | InterruptedException e) {
                LOG.trace("Writer disconnected from node " + node + ": " + e.getMessage());
                disconnect();
            }
        }
    }
}

File: networking/src/main/kotlin/ch/dissem/bitmessage/networking/nio/Connection.kt
/*
 * Copyright 2017 Christian Basler
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package ch.dissem.bitmessage.networking.nio

import ch.dissem.bitmessage.InternalContext
import ch.dissem.bitmessage.InternalContext.Companion.NETWORK_EXTRA_BYTES
import ch.dissem.bitmessage.InternalContext.Companion.NETWORK_NONCE_TRIALS_PER_BYTE
import ch.dissem.bitmessage.entity.*
import ch.dissem.bitmessage.entity.valueobject.InventoryVector
import ch.dissem.bitmessage.entity.valueobject.NetworkAddress
import ch.dissem.bitmessage.exception.InsufficientProofOfWorkException
import ch.dissem.bitmessage.ports.NetworkHandler
import ch.dissem.bitmessage.utils.Singleton.cryptography
import ch.dissem.bitmessage.utils.UnixTime
import ch.dissem.bitmessage.utils.UnixTime.MINUTE
import org.slf4j.LoggerFactory
import java.io.IOException
import java.util.*
import java.util.concurrent.ConcurrentHashMap

/**
 * Contains everything used by both the old streams-oriented NetworkHandler and the new NioNetworkHandler,
 * respectively their connection objects.
 */
class Connection(
    private val ctx: InternalContext,
    val mode: Mode,
    val node: NetworkAddress,
    private val commonRequestedObjects: MutableMap<InventoryVector, Long>,
    syncTimeout: Long
) {
    private val requestedObjects: MutableSet<InventoryVector> = Collections.newSetFromMap(ConcurrentHashMap<InventoryVector, Boolean>(10000))

    internal val io = ConnectionIO(mode, syncTimeout, commonRequestedObjects, requestedObjects, { state }, this::handleMessage)
    private var initializer: NetworkConnectionInitializer? = NetworkConnectionInitializer(ctx, node, mode, io::send) { s ->
        state = State.ACTIVE
        streams = s
        initializer = null
    }

    private val listener: NetworkHandler.MessageListener = ctx.networkListener
    private val ivCache: MutableMap<InventoryVector, Long> = ConcurrentHashMap()

    private var lastObjectTime: Long = 0

    lateinit var streams: LongArray
        protected set

    @Volatile var state = State.CONNECTING
        private set

    val isSyncFinished
        get() = io.isSyncFinished

    val nothingToSend
        get() = io.sendingQueue.isEmpty()

    init {
        initializer!!.start()
    }

    fun send(payload: MessagePayload) = io.send(payload)

    protected fun handleMessage(payload: MessagePayload) {
        when (state) {
            State.CONNECTING -> initializer!!.handleCommand(payload)
            State.ACTIVE -> receiveMessage(payload)
            State.DISCONNECTED -> disconnect()
        }
    }

    private fun receiveMessage(messagePayload: MessagePayload) {
        when (messagePayload.command) {
            MessagePayload.Command.INV -> receiveMessage(messagePayload as Inv)
            MessagePayload.Command.GETDATA -> receiveMessage(messagePayload as GetData)
            MessagePayload.Command.OBJECT -> receiveMessage(messagePayload as ObjectMessage)
            MessagePayload.Command.ADDR -> receiveMessage(messagePayload as Addr)
            else -> throw IllegalStateException("Unexpectedly received '${messagePayload.command}' command")
        }
    }

    private fun receiveMessage(inv: Inv) {
        val originalSize = inv.inventory.size
        updateIvCache(inv.inventory)
        val missing = ctx.inventory.getMissing(inv.inventory, *streams)
        LOG.trace("Received inventory with $originalSize elements, of which are ${missing.size} missing.")
        io.send(GetData(missing - commonRequestedObjects.keys))
    }

    private fun receiveMessage(getData: GetData) {
        getData.inventory.forEach { iv -> ctx.inventory.getObject(iv)?.let { obj -> io.send(obj) } }
    }

    private fun receiveMessage(objectMessage: ObjectMessage) {
        requestedObjects.remove(objectMessage.inventoryVector)
        if (ctx.inventory.contains(objectMessage)) {
            LOG.trace("Received object " + objectMessage.inventoryVector + " - already in inventory")
            return
        }
        try {
            listener.receive(objectMessage)
            cryptography().checkProofOfWork(objectMessage, NETWORK_NONCE_TRIALS_PER_BYTE, NETWORK_EXTRA_BYTES)
            ctx.inventory.storeObject(objectMessage)
            // offer object to some random nodes so it gets distributed throughout the network:
            ctx.networkHandler.offer(objectMessage.inventoryVector)
            lastObjectTime = UnixTime.now
        } catch (e: InsufficientProofOfWorkException) {
            LOG.warn(e.message)
            // DebugUtils.saveToFile(objectMessage); // this line must not be committed active
        } catch (e: IOException) {
            LOG.error("Stream " + objectMessage.stream + ", object type " + objectMessage.type + ": " + e.message, e)
        } finally {
            if (commonRequestedObjects.remove(objectMessage.inventoryVector) == null) {
                LOG.debug("Received object that wasn't requested.")
            }
        }
    }

    private fun receiveMessage(addr: Addr) {
        LOG.trace("Received " + addr.addresses.size + " addresses.")
        ctx.nodeRegistry.offerAddresses(addr.addresses)
    }

    private fun updateIvCache(inventory: List<InventoryVector>) {
        cleanupIvCache()
        val now = UnixTime.now
        for (iv in inventory) {
            ivCache.put(iv, now)
        }
    }

    fun offer(iv: InventoryVector) {
        io.send(Inv(listOf(iv)))
        updateIvCache(listOf(iv))
    }

    fun knowsOf(iv: InventoryVector): Boolean {
        return ivCache.containsKey(iv)
    }

    fun requested(iv: InventoryVector): Boolean {
        return requestedObjects.contains(iv)
    }

    private fun cleanupIvCache() {
        val fiveMinutesAgo = UnixTime.now - 5 * MINUTE
        for ((key, value) in ivCache) {
            if (value < fiveMinutesAgo) {
                ivCache.remove(key)
            }
        }
    }

    // the TCP timeout starts out at 20 seconds
    // after verack messages are exchanged, the timeout is raised to 10 minutes
    fun isExpired(): Boolean = when (state) {
        State.CONNECTING -> io.lastUpdate < System.currentTimeMillis() - 20000
        State.ACTIVE -> io.lastUpdate < System.currentTimeMillis() - 600000
        State.DISCONNECTED -> true
    }

    fun disconnect() {
        state = State.DISCONNECTED
        io.disconnect()
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is Connection) return false
        return node == other.node
    }

    override fun hashCode(): Int {
        return Objects.hash(node)
    }

    enum class State {
        CONNECTING, ACTIVE, DISCONNECTED
    }

    enum class Mode {
        SERVER, CLIENT, SYNC
    }

    companion object {
        private val LOG = LoggerFactory.getLogger(Connection::class.java)
    }
}
